opal/util/json/opal_json.h

```diff
@@ -0,0 +1,165 @@
+     goto out;
+}
+
+ret = opal_json_get_key(json, "key", &item); // Get the item pointed to by "key"
+if (OPAL_SUCCESS != ret) {
+     goto out;
+}
+
+ret = opal_json_read_string(item, (const char **) &val, &len); // Read the string value and length
+if (OPAL_SUCCESS != ret) {
+    goto out;
+}
+
+out:
+
+// Cleanup
+opal_json_free(&json);
+opal_json_free(&item);
+
+ */
+
+#ifndef OPAL_JSON_H
+#define OPAL_JSON_H
+
+#include "opal/mca/mca.h"
+
+BEGIN_C_DECLS
+
+typedef enum {
+    OPAL_JSON_NULL,
+    OPAL_JSON_OBJECT,
+    OPAL_JSON_ARRAY,
+    OPAL_JSON_INTEGER,
+    OPAL_JSON_DOUBLE,
+    OPAL_JSON_BOOL,
+    OPAL_JSON_STRING,
+    OPAL_JSON_TYPE_COUNT,
+} opal_json_type;
+
+struct opal_json_t {
+    opal_json_type type;
+};
+typedef struct opal_json_t opal_json_t;
+
+/**
+ * Load JSON from a string.
+ *
+ * @param[in]   str     A JSON string
+ * @param[in]   len     First bytes to parse
+ * @param[out]  json    Output JSON object if the string is valid
+ *
+ * @returns     OPAL_SUCCESS if the string is parsed successfully
+ *              OPAL_ERROR otherwise
+ */
+OPAL_DECLSPEC int opal_json_load(const char *str, const size_t len, const opal_json_t **json);
+
+/**
+ * Load JSON from a file.
+ *
+ * @param[in]   filename    File name
+ * @param[out]  json        Output JSON object if the file valid
+ *
+ * @returns     OPAL_SUCCESS if the file is read and parsed successfully
+ *              OPAL_ERROR otherwise
+ */
+OPAL_DECLSPEC int opal_json_load_file(const char *filename, const opal_json_t **json);
+
+/**
+ * Free JSON resources
+ *
+ * @param[in]  json     Pointer to JSON object to free
+ */
+OPAL_DECLSPEC void opal_json_free(const opal_json_t **json);
+
+/**
+ * Get the JSON object with a key value from a parent object
+ *
+ * @param[in]   json    Parent JSON object
+ * @param[in]   key     Key value
+ * @param[out]  out     Output JSON object with the specified key value
+ *
+ * @returns     OPAL_SUCCESS if an object is found with the specified key value
+ *              OPAL_ERROR otherwise
+ */
+OPAL_DECLSPEC int opal_json_get_key(const opal_json_t *json, const char *key,
+                                    const opal_json_t **out);
+
+/**
+ * Get the JSON object at index from a JSON array
+ *
+ * @param[in]   json    Parent JSON array
+ * @param[in]   index   Index value
+ * @param[out]  out     Output JSON object at the specified index
+ *
+ * @returns     OPAL_SUCCESS if an object is found at the specified index
+ *              OPAL_ERROR otherwise
+ */
+OPAL_DECLSPEC int opal_json_get_index(const opal_json_t *json, const size_t index,
+                                      const opal_json_t **out);
+
+/**
+ * Get the JSON object size
+ *
+ * @param[in]   json    A JSON object
+ * @param[out]  size    Number of elements in the object
+ *
+ * @returns     OPAL_SUCCESS if successful
+ *              OPAL_ERROR otherwise
+ */
+OPAL_DECLSPEC int opal_json_get_object_size(const opal_json_t *json, size_t *size);
```

Comment:
Is this returning the number of elements in the json object, or the size of the object in bytes?

The name is a little misleading -- IMHO, `size` refers to number of bytes.  Could we use `opal_json_get_num_keys` or somesuch?

Reply:
An object is untyped, it cannot contain elements, it only contains bytes. The other function you are referring to is `opal_json_get_array_len` but it can only apply to arrays.

Reply:
@bosilca Actually this function is also counting object length, not bytes.

I wonder if I should simply keep 1 function for both object and array, i.e. `opal_json_get_len`.

Reply:
got it, object is a keymapped values storage of `json_value_t*` while array is just an array of `json_value_t*`. Then a single assessor would be nice.

Reply:
Thanks. I decided to keep a single function `opal_json_get_container_size` for both array and object. They are too similar to have different functions.